### ISSUE TITLE
papis rename does not rename the PDF files when add-file-name is set

### ISSUE THREAD

**Issue by stylerw** [papis/papis#894]
* papis version (`$ papis --version` or commit number): 129773d

**Expected Behavior**:

Running `papis rename -a` will change the folder names, and filenames of associated PDFs, when run, according to the rules in .config for add-folder-name and add-file-name as below

```
add-file-name = {doc[author]}_{doc[year]}_{doc[title]}
add-folder-name = {doc[author]}_{doc[year]}
```

**Observed Behavior**:

Running `papis rename -ab` will change the folder names, but filenames will not be updated to match the set standard

**Why this is a problem**: Many of us use tools like fzf to find files rapidly, and semantic filenames are useful. It is also silly to have to manually change the filename when a bibtex entry exists which has the information I'd like.

Thanks!

**Comment by pmiam** [comment]
I have also been having problems with `papis rename` in structured library I recently set up for a long-term project. I'll try to take a look at this soon.

**Comment by klb2** [comment]
Hi,
Are there any updates on this issue or a temporary solution?

**Comment by alexfikl** [comment]
> Hi, Are there any updates on this issue or a temporary solution?

There's some talk in #1104 to merge the move and rename commands that should include this bug. It's still early stages (as you can probably see :grin:)

**Comment by klb2** [comment]
Great thank you. Looking forward to this

**Comment by jghauser** [comment]
Just a quick comment on this. While I've made a fair bit of progress on #1104, this will not actually fix the issue here. The proposal at the moment is to improve `papis update` to rename files instead. @pmiam [once](https://github.com/papis/papis/pull/826#issuecomment-2467159515) expressed interest in working towards that, but I'm not sure whether they're still interested in it. I might get to it once I finish the work on `papis mv`. Eventually, we might also want to have a flag in `papis mv` to trigger a filename update with `papis update`.

**Comment by alexfikl** [comment]
> Just a quick comment on this. While I've made a fair bit of progress on [#1104](https://github.com/papis/papis/pull/1104), this will not actually fix the issue here.

Oops, sorry for the confusion!

> I might get to it once I finish the work on papis mv. Eventually, we might also want to have a flag in papis mv to trigger a filename update with papis update.

This fell off my radar as well, but it sounds rather simple to modify `papis update` to rename files as mentioned, so I might give it a go over the weekend. :grin:

**Comment by jghauser** [comment]
[Here](https://github.com/papis/papis/pull/826#issuecomment-2467717721) are some notes I made about this at the time. I guess the first part about renaming folders doesn't apply. Separation of concerns isn't quite clear as the `info.yaml` could be seen as a source of truth for files and folders and arguably we might argue that `papis update` should lead to updates of both. But at the moment I think maybe the clearest would be to have `papis update` deal with anything that is explicitly set in the `info.yaml` (incl `files` and `notes`) and making sure files on disk are named correspondingly. `papis mv` then deals with folder names (which aren't explicitly set in the `info.yaml`).

> This fell off my radar as well, but it sounds rather simple to modify `papis update` to rename files as mentioned, so I might give it a go over the weekend. 😁

Looking forward ❤️🎉

**Comment by alexfikl** [comment]
> [Here](https://github.com/papis/papis/pull/826#issuecomment-2467717721) are some notes I made about this at the time. 

Thanks! It's getting a bit difficult to keep track of these older discussions :(

> But at the moment I think maybe the clearest would be to have `papis update` deal with anything that is explicitly set in the `info.yaml` (incl `files` and `notes`) and making sure files on disk are named correspondingly. `papis mv` then deals with folder names (which aren't explicitly set in the `info.yaml`).

That sounds like a conceptually intuitive way to separate the two, so I agree :grin:

**Comment by klb2** [comment]
> [Here](https://github.com/papis/papis/pull/826#issuecomment-2467717721) are some notes I made about this at the time. I guess the first part about renaming folders doesn't apply. Separation of concerns isn't quite clear as the `info.yaml` could be seen as a source of truth for files and folders and arguably we might argue that `papis update` should lead to updates of both. But at the moment I think maybe the clearest would be to have `papis update` deal with anything that is explicitly set in the `info.yaml` (incl `files` and `notes`) and making sure files on disk are named correspondingly. `papis mv` then deals with folder names (which aren't explicitly set in the `info.yaml`).

I only started using papis very recently and having `info.yaml` as the source of truth for files and folder is what I expected.

In particular, I ran into an issue when I imported some files using their DOIs. Some of the information was incorrect or could not be imported. This caused some wrong folder/file names. So I manually corrected the errors in the info.yaml file and expected the `papis update` command to automatically rename the folder and files (according to `add-file-name` and `add-folder-name`) using the updated information in `info.yaml`.

**Comment by jghauser** [comment]
Actually, while looking at this I noticed another problem. I don't think `papis update` has a way of setting a value to the default defined in e.g. `add-file-name`. Maybe we can change `papis update --set` so that it renames to the default (if it exists) when called with only one argument (e.g. `papis update --set`). We also need to keep track of which file is which and handle both lists (e.g. `files`) and strings (e.g. `notes`). So yeah, it might not be all that simple 😉.

**Comment by jghauser** [comment]
> I only started using papis very recently and having `info.yaml` as the source of truth for files and folder is what I expected.
> 
> In particular, I ran into an issue when I imported some files using their DOIs. Some of the information was incorrect or could not be imported. So I manually corrected the errors in the info.yaml file and expected the `papis update` command to automatically rename the folder and files (according to `add-file-name` and `add-folder-name`) using the updated information in `info.yaml`.

Yeah, I think that assumption makes sense. I also like a very automatic handling of files and folder names (using the `info.yaml` as the source of truth. Others might do it differently however and use `papis mv` to move docs around. I think the separation between `papis mv` and `papis update` as mentioned above is the cleanest division that accommodates different use cases. This would require running `papis update` and `papis mv` both to achieve what you and I usually want. We might eventually streamline this with a flag and/or config option though. I'm open to other suggestions though!

**Comment by alexfikl** [comment]
> Actually, while looking at this I noticed another problem. I don't think `papis update` has a way of setting a value to the default defined in e.g. `add-file-name`. Maybe we can change `papis update --set` so that it renames to the default (if it exists) when called with only one argument (e.g. `papis update --set`). We also need to keep track of which file is which and handle both lists (e.g. `files`) and strings (e.g. `notes`). So yeah, it might not be all that simple 😉.

I'm perfectly fine with special casing `files` and `notes` in there. At least for now, I don't know if it needs a more general solution?

> So I manually corrected the errors in the info.yaml file and expected the papis update command to automatically rename the folder and files (according to add-file-name and add-folder-name) using the updated information in info.yaml.

Agreed as well! I actually have a personal "papis clean" command that does a bunch of cleanup (including renaming files and folders), so I got lazy about actually fixing it in papis :(

**Comment by jghauser** [comment]
> > Actually, while looking at this I noticed another problem. I don't think `papis update` has a way of setting a value to the default defined in e.g. `add-file-name`. Maybe we can change `papis update --set` so that it renames to the default (if it exists) when called with only one argument (e.g. `papis update --set`). We also need to keep track of which file is which and handle both lists (e.g. `files`) and strings (e.g. `notes`). So yeah, it might not be all that simple 😉.
> 
> I'm perfectly fine with special casing `files` and `notes` in there. At least for now, I don't know if it needs a more general solution?

Yeah, special cases for `files` and `notes` makes sense. But even then they could be changed with `papis update --set` or the various `papis update` arguments that change lists (in the case of `files`). I mean I also don't understand why `notes` is a string and not just a set as well, but that's only opening another can of worms...

> > So I manually corrected the errors in the info.yaml file and expected the papis update command to automatically rename the folder and files (according to add-file-name and add-folder-name) using the updated information in info.yaml.
> 
> Agreed as well! I actually have a personal "papis clean" command that does a bunch of cleanup (including renaming files and folders), so I got lazy about actually fixing it in papis :(

Such a command would be very useful! Maybe we can somehow eventually integrate it into `papis update`.

**Comment by pmiam** [comment]
Please excuse my absence. I didn't ever make much progress on this. I
was side-tracked by the allure of much more ambitious UI redesign, and
then I got a job that's kept me busy since - haha.

With a pragmatic approach, I think this feature can be hacked together
easily. Here is a zsh alias I used successfully a long time ago (I no
longer use it).

```
alias papis='papis '
alias refgen='update -s ref.jinja2 "$(papis config -s lit --json | jq ".\"ref-format.jinja2\"")"'
```

This is invoked in zsh as `papis refgen`. It was meant to recreate
malformed ref keys in case I inducted a document without all the
information expected by my complicated preferences encoded as a heavily
customized jinja template somewhere in my (unfortunately also outdated)
dotfiles.

With a few more pipes and maybe a tmp file, I think this kind of thing
could handle renaming files. Of course, this is far from ideal.
